### PR TITLE
Allow Razor generated documents to respect tabs/spaces settings

### PR DESCRIPTION
**Pull request by allisonchou**
Part of https://github.com/dotnet/aspnetcore-tooling/pull/3317. Enables generated documents to respect the tabs/spaces settings passed in from Razor tooling.

The bulk of this is Ajay's work from https://github.com/dotnet/aspnetcore-tooling/pull/1606, so major creds to him. 🎉

### CODE REVIEW

**Review by TanayParikh — APPROVED** [review]
LGTM!

File: src/Razor/Microsoft.AspNetCore.Razor.Language/src/CodeGeneration/CodeWriter.cs [hunk 1]
```diff
@@ -62,16 +72,43 @@ public string NewLine
...
+            if (IndentWithTabs)
+            {
+                // Avoid writing directly to the StringBuilder here, that will throw off the manual indexing 
+                // done by the base class.
+                var tabs = size / TabSize;
+                actualSize += tabs;
+                for (var i = 0; i < tabs; i++)
+                {
+                    _builder.Append("\t");
+                }
```
Comment (campersau) [inline src/Razor/Microsoft.AspNetCore.Razor.Language/src/CodeGeneration/CodeWriter.cs:92]: Could be simplified to:
```suggestion
                _builder.Append('\t', tabs);
```
(also in the two places below)

Comment (allisonchou) [reply src/Razor/Microsoft.AspNetCore.Razor.Language/src/CodeGeneration/CodeWriter.cs]: Thanks! Updated.

File: src/Razor/Microsoft.AspNetCore.Razor.Language/test/TestFiles/IntegrationTests/ComponentDesignTimeCodeGenerationTest/SingleLineControlFlowStatements_InCodeBlock/TestComponent.codegen.cs [hunk 2]
```diff
@@ -45,7 +45,7 @@ protected override void BuildRenderTree(Microsoft.AspNetCore.Components.Renderin
 #line default
 #line hidden
 #nullable disable
-                                      
+
```
Comment (allisonchou) [inline src/Razor/Microsoft.AspNetCore.Razor.Language/test/TestFiles/IntegrationTests/ComponentDesignTimeCodeGenerationTest/SingleLineControlFlowStatements_InCodeBlock/TestComponent.codegen.cs:48]: @TanayParikh It looks like Ajay was seeing the same issue in https://github.com/dotnet/aspnetcore-tooling/pull/1606. It doesn't look too impactful, but just want to check this change is OK?

File: src/Razor/Microsoft.AspNetCore.Razor.Language/src/CodeGeneration/CodeWriter.cs [hunk 3]
```diff
@@ -18,9 +18,15 @@ public sealed class CodeWriter
         private int _currentLineIndex;
         private int _currentLineCharacterIndex;
 
-        public CodeWriter()
+        public CodeWriter() : this(Environment.NewLine, false, 4)
```
Comment (pranavkm) [inline src/Razor/Microsoft.AspNetCore.Razor.Language/src/CodeGeneration/CodeWriter.cs:21]:
```suggestion
        public CodeWriter() : this(Environment.NewLine, indentWithTabs: false, tabSize: 4)
```

File: src/Razor/Microsoft.AspNetCore.Razor.Language/src/CodeGeneration/CodeWriter.cs [hunk 4]
```diff
@@ -62,16 +72,34 @@ public string NewLine
             }
         }
 
-        // Internal for testing.
-        internal CodeWriter Indent(int size)
+        public CodeWriter Indent(int size)
         {
-            if (Length == 0 || this[Length - 1] == '\n')
+            if (size == 0 || (Length != 0 && this[Length - 1] != '\n'))
```
Comment (pranavkm) [inline src/Razor/Microsoft.AspNetCore.Razor.Language/src/CodeGeneration/CodeWriter.cs:77]: This is unrelated to your PR, but indexing in to `StringBuilder` (which is what `this[..]` is doing) is really slow.

**Review by pranavkm — APPROVED** [review]

File: src/Razor/Microsoft.AspNetCore.Razor.Language/src/CodeGeneration/CodeWriter.cs (hunk 4 above)
Comment (allisonchou) [reply src/Razor/Microsoft.AspNetCore.Razor.Language/src/CodeGeneration/CodeWriter.cs:77]: Hmm, interesting. Maybe we can keep track of the last character somehow, possibly storing it in a field?
(Don't know if it's in the scope of this PR, but would be a good follow up later on.)

File: src/Razor/Microsoft.AspNetCore.Razor.Language/src/CodeGeneration/DefaultDocumentWriter.cs [hunk 5]
```diff
@@ -32,7 +32,7 @@ public override RazorCSharpDocument WriteDocument(RazorCodeDocument codeDocument
             }
 
             var context = new DefaultCodeRenderingContext(
-                new CodeWriter(),
+                new CodeWriter(Environment.NewLine, _options.IndentWithTabs, _options.IndentSize),
```
Comment (NTaylorMullen) [inline src/Razor/Microsoft.AspNetCore.Razor.Language/src/CodeGeneration/DefaultDocumentWriter.cs:35]: Hmmm I wonder if it makes more sense for the code writer to take in the code gen options instead. @pranavkm thoughts now that you own this? 😄

File: src/Razor/Microsoft.AspNetCore.Razor.Language/test/TestFiles/IntegrationTests/ComponentDesignTimeCodeGenerationTest/SingleLineControlFlowStatements_InCodeBlock/TestComponent.codegen.cs (hunk 2 above)
Comment (NTaylorMullen) [reply src/Razor/Microsoft.AspNetCore.Razor.Language/test/TestFiles/IntegrationTests/ComponentDesignTimeCodeGenerationTest/SingleLineControlFlowStatements_InCodeBlock/TestComponent.codegen.cs:48]: Hmmm, so this wont have any actual impact on the runnability of the code since it's outside of the line pragma; however, it's concerning that this is being touched at all? Definitely need to look into what's happening here. Also, in the other files I wouldn't imagine that generated locations would change unless there was a mashup of tabs in the source files / interesting settings in the tests.

**Review by NTaylorMullen — CHANGES_REQUESTED** [review]
Looking good! One big concern is that the generated files changed. Need to validate that it's expected and that we were doing the wrong thing previously and not doing the wrong thing now. I wouldn't have imagined they'd change is all

File: src/Razor/Microsoft.AspNetCore.Razor.Language/test/TestFiles/IntegrationTests/ComponentDesignTimeCodeGenerationTest/SingleLineControlFlowStatements_InCodeBlock/TestComponent.codegen.cs (hunk 2 above)
Comment (NTaylorMullen) [reply src/Razor/Microsoft.AspNetCore.Razor.Language/test/TestFiles/IntegrationTests/ComponentDesignTimeCodeGenerationTest/SingleLineControlFlowStatements_InCodeBlock/TestComponent.codegen.cs:48]: Also, @pranavkm does this not worry you?

**Comment by allisonchou** [comment]
Okay, so I think I see what's going on here in regards to why the code gen and mappings have changed.
tl;dr I think the current changes are correct.

In the existing code, the indent logic for padding is done [here](https://github.com/dotnet/aspnetcore/blob/b00ae1ba5f69db493d4187885b5f01fb48698605/src/Razor/Microsoft.AspNetCore.Razor.Language/src/CodeGeneration/CodeWriterExtensions.cs#L51) in CodeWriterExtensions. However, I believe this logic is incorrect since it does not properly increment the[ `_currentLineCharacterIndex` and `_absoluteIndex` variables](https://github.com/dotnet/aspnetcore/blob/b00ae1ba5f69db493d4187885b5f01fb48698605/src/Razor/Microsoft.AspNetCore.Razor.Language/src/CodeGeneration/CodeWriter.cs#L18) in the backing CodeWriter. The current PR correctly increments the variables, hence the changes.

cc @NTaylorMullen

File: src/Razor/Microsoft.AspNetCore.Razor.Language/src/CodeGeneration/DefaultDocumentWriter.cs (hunk 5 above)
Comment (allisonchou) [reply src/Razor/Microsoft.AspNetCore.Razor.Language/src/CodeGeneration/DefaultDocumentWriter.cs]: I think you're right, it would also make it easier in the future if we were to add other code gen option variables to the constructor.

**Comment by NTaylorMullen** [comment]
> Okay, so I think I see what's going on here in regards to why the code gen and mappings have changed.
> tl;dr I think the current changes are correct.
> 
> In the existing code, the indent logic for padding is done [here](https://github.com/dotnet/aspnetcore/blob/b00ae1ba5f69db493d4187885b5f01fb48698605/src/Razor/Microsoft.AspNetCore.Razor.Language/src/CodeGeneration/CodeWriterExtensions.cs#L51) in CodeWriterExtensions. However, I believe this logic is incorrect since it does not properly increment the[ `_currentLineCharacterIndex` and `_absoluteIndex` variables](https://github.com/dotnet/aspnetcore/blob/b00ae1ba5f69db493d4187885b5f01fb48698605/src/Razor/Microsoft.AspNetCore.Razor.Language/src/CodeGeneration/CodeWriter.cs#L18) in the backing CodeWriter. The current PR correctly increments the variables, hence the changes.
> 
> cc @NTaylorMullen

Ahhhh I can totally see that. Thank you for validating @allisonchou !

**Review by NTaylorMullen — APPROVED** [review]

**Review by ryanbrandenburg — APPROVED** [review]

**Comment by ryanbrandenburg** [comment]
The failure looks random, will probably pass on retry.

**Comment by BrennanConroy** [comment]
> The failure looks random, will probably pass on retry.

Or, you could file an issue and mark the test as quarantined.

**Comment by BrennanConroy** [comment]
https://github.com/dotnet/aspnetcore/pull/31405
Try not to ignore test failures
